lib/rack/lint.rb

```diff
@@ -673,19 +673,23 @@ def check_headers(headers)
           ## Header keys must not contain uppercase ASCII characters (A-Z).
           raise LintError, "uppercase character in header name: #{key}" if key =~ /[A-Z]/
 
-          ## The values of the header must be Strings,
-          unless value.kind_of? String
-            raise LintError, "a header value must be a String, but the value of '#{key}' is a #{value.class}"
-          end
-          ## consisting of lines (for multiple header values, e.g. multiple
-          ## <tt>Set-Cookie</tt> values) separated by "\\n".
-          value.split("\n").each { |item|
-            ## The lines must not contain characters below 037.
-            if item =~ /[\000-\037]/
-              raise LintError, "invalid header value #{key}: #{item.inspect}"
-            end
-          }
-        }
+          ## Header values must be either a String instance,
+          if value.kind_of?(String)
+            check_header_value(key, value)
+          elsif value.kind_of?(Array)
+            ## or an Array of String instances,
+            value.each{|value| check_header_value(key, value)}
+          else
+            raise LintError, "a header value must be a String or Array of Strings, but the value of '#{key}' is a #{value.class}"
+          end
+        end
+      end
+
+      def check_header_value(key, value)
+        ## such that each String instance must not contain characters below 037.
```

Comment:
If servers want to support both Rack 1/2 and Rack 3 applications, they'll need to still check for newlines in strings.  As newlines in strings are not valid in Rack 3, the server can assume if they are present, they are dealing with Rack 1/2.  This isn't a problem, I just want to confirm that's how we expect servers to handle this.

Reply:
I think that’s fair, what are the implications?

Reply:
I guess that the implications for servers would be that in case `\n` was accidentally included in the header value, instead of skipping or truncating the header, we would end up with multiple headers (per the Rack 2 protocol).